File: app/src/main/java/com/dpis/module/templates/TemplateTypefaceResolver.java
package com.dpis.module.templates;

import com.dpis.module.fonts.SystemFontEntry;
import com.dpis.module.fonts.SystemFontRegistry;

public final class TemplateTypefaceResolver implements TemplateConfigSummaryFormatter.TypefaceResolver {
    public interface ImportedTypefaceProvider {
        TemplateConfigSummaryFormatter.TypefaceStatus resolve(String typefaceId);
    }

    public interface SystemTypefaceProvider {
        boolean canLoad(String typefaceId);

        String displayName(String typefaceId);
    }

    private final ImportedTypefaceProvider importedTypefaceProvider;
    private final SystemTypefaceProvider systemTypefaceProvider;

    public TemplateTypefaceResolver(ImportedTypefaceProvider importedTypefaceProvider) {
        this(importedTypefaceProvider, new AndroidSystemTypefaceProvider());
    }

    public TemplateTypefaceResolver(ImportedTypefaceProvider importedTypefaceProvider,
            SystemTypefaceProvider systemTypefaceProvider) {
        this.importedTypefaceProvider = importedTypefaceProvider;
        this.systemTypefaceProvider = systemTypefaceProvider != null
                ? systemTypefaceProvider
                : new AndroidSystemTypefaceProvider();
    }

    @Override
    public TemplateConfigSummaryFormatter.TypefaceStatus resolve(String typefaceId) {
        if (typefaceId == null || typefaceId.isBlank()) {
            return TemplateConfigSummaryFormatter.TypefaceStatus.none();
        }
        if (SystemFontRegistry.isSystemFontId(typefaceId)) {
            if (systemTypefaceProvider.canLoad(typefaceId)) {
                return TemplateConfigSummaryFormatter.TypefaceStatus.resolved(
                        typefaceId, systemTypefaceProvider.displayName(typefaceId));
            }
            return TemplateConfigSummaryFormatter.TypefaceStatus.missing(typefaceId);
        }
        if (importedTypefaceProvider != null) {
            TemplateConfigSummaryFormatter.TypefaceStatus imported =
                    importedTypefaceProvider.resolve(typefaceId);
            if (imported != null && imported.resolved()) {
                return imported;
            }
        }
        return TemplateConfigSummaryFormatter.TypefaceStatus.missing(typefaceId);
    }

    private static final class AndroidSystemTypefaceProvider implements SystemTypefaceProvider {
        @Override
        public boolean canLoad(String typefaceId) {
            return SystemFontRegistry.loadTypeface(typefaceId) != null;
        }

        @Override
        public String displayName(String typefaceId) {
            for (SystemFontEntry entry : SystemFontRegistry.listRecommendedFonts()) {
                if (typefaceId.equals(entry.id())) {
                    return entry.displayName();
                }
            }
            return typefaceId;
        }
    }
}

File: app/src/main/java/com/dpis/module/templates/TemplateTypefaceResolver.kt
package com.dpis.module.templates

import com.dpis.module.fonts.SystemFontRegistry
import com.dpis.module.templates.TemplateConfigSummaryFormatter.TypefaceResolver
import com.dpis.module.templates.TemplateConfigSummaryFormatter.TypefaceStatus

class TemplateTypefaceResolver @JvmOverloads constructor(
    private val importedTypefaceProvider: ImportedTypefaceProvider?,
    systemTypefaceProvider: SystemTypefaceProvider? = AndroidSystemTypefaceProvider()
) : TypefaceResolver {
    fun interface ImportedTypefaceProvider {
        fun resolve(typefaceId: String?): TypefaceStatus?
    }

    interface SystemTypefaceProvider {
        fun canLoad(typefaceId: String?): Boolean

        fun displayName(typefaceId: String?): String?
    }

    private val systemTypefaceProvider: SystemTypefaceProvider

    init {
        this.systemTypefaceProvider = if (systemTypefaceProvider != null)
            systemTypefaceProvider
        else
            AndroidSystemTypefaceProvider()
    }

    override fun resolve(typefaceId: String?): TypefaceStatus {
        if (typefaceId == null || typefaceId.isBlank()) {
            return TypefaceStatus.none()
        }
        if (SystemFontRegistry.isSystemFontId(typefaceId)) {
            if (systemTypefaceProvider.canLoad(typefaceId)) {
                return TypefaceStatus.resolved(
                    typefaceId, systemTypefaceProvider.displayName(typefaceId)
                )
            }
            return TypefaceStatus.absent(typefaceId)
        }
        if (importedTypefaceProvider != null) {
            val imported =
                importedTypefaceProvider.resolve(typefaceId)
            if (imported != null && imported.resolved()) {
                return imported
            }
        }
        return TypefaceStatus.absent(typefaceId)
    }

    private class AndroidSystemTypefaceProvider : SystemTypefaceProvider {
        override fun canLoad(typefaceId: String?): Boolean {
            return SystemFontRegistry.loadTypeface(typefaceId) != null
        }

        override fun displayName(typefaceId: String?): String? {
            if (typefaceId == null) return null
            for (entry in SystemFontRegistry.listRecommendedFonts()) {
                if (typefaceId == entry.id()) {
                    return entry.displayName()
                }
            }
            return typefaceId
        }
    }
}
